### DISCUSSION TITLE
custom illegal opcode execution?

### DISCUSSION THREAD

**Discussion by RR7** [dirkwhoffmann/Moira#26] [Q&A] [answered]
what's the 'correct' process for this? i inject an illegal opcode, handle it with willExecute, but then moira continues to process the exception. i can edit the moira source code to not do this, but i'm wondering if i'm missing something? i'd like to keep moira untouched of possible.

**Comment by dirkwhoffmann** [comment]
Moira has a feature called *software traps*, but I am not sure if it fits your needs. A software trap pauses emulation when a certain instruction is executed. When a software trap is set at some address, the instruction is remembered and replaced by a specific LINE-A instruction. When this LINE-A instruction is executed, Moira pauses emulation, calls the delegate `softwareTrapReached`, and restores the original instruction in the prefetch queue. Thus, Moira executes the original code rather than the exception. 

Software traps are handles in this function:

```c++
template <Core C, Instr I, Mode M, Size S> void
Moira::execLineA(u16 opcode)
{
    AVAILABILITY(C68000)

    // Check if a software trap is set for this instruction
    if (debugger.swTraps.traps.contains(opcode)) {

        auto &trap = debugger.swTraps.traps[opcode];

        // Restore the original instruction
        reg.pc = reg.pc0;
        queue.irc = trap.instruction;
        prefetch<C>();

        // Inform the delegate
        softwareTrapReached(reg.pc0);
        return;
    }
    ...
}
``` 

In vAmiga, this feature is used to let us pause emulation when an application starts to execute. For doing so, it patches an ADF file with a software trap at the proper location of an executable. It is best explained by showing some code:

```c++
void
FloppyDrive::catchFile(const std::filesystem::path &path)
{
    {   SUSPENDED
        
        // Extract the file system
        auto fs = MutableFileSystem(*this);
        
        // Seek file
        auto file = fs.seekFile(path.string());
        if (file == nullptr) throw Error(VAERROR_FILE_NOT_FOUND);
        
        // Extract file
        Buffer<u8> buffer;
        file->writeData(buffer);
        
        // Parse hunks
        auto descr = ProgramUnitDescriptor(buffer);
        
        // Seek the code section and read the first instruction word
        auto offset = descr.seek(HUNK_CODE);
        if (!offset) throw Error(VAERROR_HUNK_CORRUPTED);
        u16 instr = HI_LO(buffer[*offset + 8], buffer[*offset + 9]);
        
        // Replace the first instruction word by a software trap
        auto trap = cpu.debugger.swTraps.create(instr);
        buffer[*offset + 8] = HI_BYTE(trap);
        buffer[*offset + 9] = LO_BYTE(trap);
        
        // Write the modification back to the file system
        file->overwriteData(buffer);
        
        // Convert the modified file system back to a disk
        auto adf = ADFFile(fs);
        
        // Replace the old disk
        swapDisk(std::make_unique<FloppyDisk>(adf));
    }
}
```

**Reply by RR7** [reply]
i'm not sure. for example, musashi has a facility for exactly this, it has an illegal instruction callback with a return value, and if the callback returns anything other than 0 then the remainder of the function doesnt process, and the cpu treats the opcode as executed, not exception occurs. it looks like moira never expects an illegal opcode to be anything other than that. which is fine, but is there another way to add my own custom opcodes in that case?

**Reply by RR7** [reply]
i've been thinking more about this,  are you suggesting i use LineA opcodes for my custom ones? thus moira will let them execute as normal ones if they are trapped? what i've done so far is added my own opcodes to moira, which was not entirely pleasant as it seems quite fixed function in there. the new opcodes are basically copies of NOP, and i think use willExecute to pick up if they've been called, then i call my own code, in some instances i'm having to imcrement the PC via jump manually, which is a bit confusing, i've not yet worked out whats happening. it is sort of working though. i'm not sure how many LineA opcodes are potentially used or unused by the ST, it doesnt feel very safe.

**Comment by dirkwhoffmann** [comment]
> for example, musashi has a facility for exactly this

I see. You're referring to this option:

```c++
/* If ON, CPU will call the callback when it encounters an illegal instruction
 * passing the opcode as argument. If the callback returns 1, then it's considered
 * as a normal instruction, and the illegal exception in canceled. If it returns 0,
 * the exception occurs normally.
 * The callback looks like int callback(int opcode)
 * You should put OPT_SPECIFY_HANDLER here if you cant to use it, otherwise it will
 * use a dummy default handler and you'll have to call m68k_set_illg_instr_callback explicitely
 */
#define M68K_ILLG_HAS_CALLBACK	    OPT_OFF
#define M68K_ILLG_CALLBACK(opcode)  op_illg(opcode)
```

Seems like a usable feature to me. I could add something like a `bool lineInstrHandler(u16 opcode)` function to Moira that is called inside the handlers for line A and line F instructions. 

Then, changing the current implementation ...

```c++
template <Core C, Instr I, Mode M, Size S> void
Moira::execLineF(u16 opcode)
{
    AVAILABILITY(C68000)

    execException<C>(EXC_LINEF);

    CYCLES_68000(34)
    CYCLES_68010(38)
    CYCLES_68020(34)

    FINALIZE
}
```

to something like that should do the trick:

```c++
template <Core C, Instr I, Mode M, Size S> void
Moira::execLineF(u16 opcode)
{
    AVAILABILITY(C68000)

    if (lineInstrHandler(opcode) == false) { 

        prefetch<C>();
        return;
    }

    execException<C>(EXC_LINEF);

    CYCLES_68000(34)
    CYCLES_68010(38)
    CYCLES_68020(34)

    FINALIZE
}
```

Would it fit your needs? 

Side note: I don't want to intercept at a lower level (i.e., in execException), because this function is sometimes called in the middle of an instruction, e.g., when an invalid argument is encountered. Simply calling `prefetch` in these cases would likely put the CPU in an inconsistent state.

**Reply by RR7** [reply]
it took me a little while to try and reconcile what i'm doing with my arch nemesis, which seems to be the preferch queue. in the, this is the approach i have taken, which is yielding good results so far.

modify the willExecute exception virtual function return a value to denote if we've handled the instruction:

`int m68k::willExecute(ExceptionType exc, u16 vector) {
	switch (exc)
	{

	case EXC_ILLEGAL:
	{
		return illegalInstrucion(getIRD()); // pass the opcode in

	}
	case EXC_TRAP:
	{
		return trapInstruction(vector); // Handle traps
	}
	case EXC_LINEA:
	{
		return lineAInstruction(getIRD());
	}
	default:
	{
	break;
	}
return 0;
}`

modify the exeception handler to suit:

`Moira::execException(ExceptionType exc, int nr)
{
    u16 status = getSR();

    // Determine the exception vector number
    u16 vector = (exc == EXC_TRAP) ? u16(exc + nr) : (exc == EXC_BKPT) ? 4 : u16(exc);


    // Inform the delegate
    if (willExecute(exc, vector))
    {
        return;
    }

    // Remember the exception vector
    exception = vector;

    // Enter supervisor mode and leave trace mode
    setSupervisorMode(true);
    clearTraceFlags();

    switch (exc) {
...<continues as normal>
`

and then ensure that once i've poked the CPU, i do this: 
`moira.debugger.jump(moira.getPC());` 

this seems to be having the desired effect and remains stable. i'm not if you'd approve!

i'm not adding cycles yet, i'd prefer to do this outside of moira's internal code.

**Reply by dirkwhoffmann** [reply, marked as answer]
Since I am curious: Can you elaborate a little more about the purpose of catching illegal instructions? Are you using this to debug or implement some ST-specific features? In vAmiga, I never felt the need to customize the treatment of illegal instructions (except for the software trap case, which was used in a particular scenario).

**Reply by RR7** [reply]
sure, mostly development/super-features/cheap hacks as opposed to part of accurate ST emulation. like setting an opcode to capture GemDOS reads/writes, means you can somewhat simulate a hard disk on the host file system, which is problematic, but quite handy, or using an opcode to bring up the debugger, or tracks cycle count when an opcode is called for timing testing software. ideally, as things progress, there wont be a need for them. it seems to be working as i need it for now, although my executable is no about 5x the size, i wonder if thats due to the debugger being embedded too, or just thats how big moira is. not an issue, just an observation :)

**Comment by dirkwhoffmann** [comment]
Thanks for the explanations!

> it seems to be working as i need it for now, although my executable is no about 5x the size

I guess you mean in comparison to the earlier version, which presumably used Musashi?

The reason is, indeed, that Moira is big. I've deliberately exchanged memory consumption for speed. Moira gains its speed by heavily using templates, which causes the compiler to replicate functions multiple times. By doing so, I managed to make Moira even slightly faster than Musashi, which is already fast due to the generated exception handler. Also, Musashi has the speed advantage of pure C code, compared to Moira, which comes with a small runtime overhead due to its C++ nature.

**Reply by RR7** [reply]
the size isn't a problem, i was just curious. I've compartmentalized the CPU, so I've tested musashi, moira and an extraction of the UAE core from early versions of Hatari. this means right now moira is working within my framework, as opposed to the emulation being built around moira. musashi seems to not work correctly for reasons i couldn't find. moira doesn't not exhibit this problem, so that's great, but i am having interrupt timing issues which the UAE core doesn't suffer from, musashi behaved the same way, so it's likely my lack of understanding on the interrupt system, i may be setting it back to 0 too early, I'm not sure yet, i need to trace the behavior. but so far, I'm quite pleased that moira seems to be working. I'm also have a couple of crashes in moira that I'm not seeing in UAE core, but could be timing related just the same.
